src/logic/safe/transactions/offchainSigner/EIP712Signer.js

```diff
@@ -0,0 +1,92 @@
+// @flow
+import { EMPTY_DATA } from '~/logic/wallets/ethTransactions'
+import { getWeb3 } from '~/logic/wallets/getWeb3'
+
+const EIP712_NOT_SUPPORTED_ERROR_MSG = "EIP712 is not supported by user's wallet"
+
+const generateTypedDataFrom = async ({
+  baseGas,
+  data,
+  gasPrice,
+  gasToken,
+  nonce,
+  operation,
+  refundReceiver,
+  safeAddress,
+  safeTxGas,
+  to,
+  valueInWei,
+}) => {
+  const typedData = {
+    types: {
+      EIP712Domain: [
+        {
+          type: 'address',
+          name: 'verifyingContract',
+        },
+      ],
+      SafeTx: [
+        { type: 'address', name: 'to' },
+        { type: 'uint256', name: 'value' },
+        { type: 'bytes', name: 'data' },
+        { type: 'uint8', name: 'operation' },
+        { type: 'uint256', name: 'safeTxGas' },
+        { type: 'uint256', name: 'baseGas' },
+        { type: 'uint256', name: 'gasPrice' },
+        { type: 'address', name: 'gasToken' },
+        { type: 'address', name: 'refundReceiver' },
+        { type: 'uint256', name: 'nonce' },
+      ],
+    },
+    domain: {
+      verifyingContract: safeAddress,
+    },
+    primaryType: 'SafeTx',
+    message: {
+      to,
+      value: valueInWei,
+      data,
+      operation,
+      safeTxGas,
+      baseGas,
+      gasPrice,
+      gasToken,
+      refundReceiver,
+      nonce: Number(nonce),
+    },
+  }
+
+  return typedData
+}
+
+type EIP712RpcCallVersion = 'v3'
+
+export const getEIP712Signer = (version: ?EIP712RpcCallVersion) => async txArgs => {
+  const web3 = getWeb3()
+  const typedData = await generateTypedDataFrom(txArgs)
+
+  const jsonTypedData = JSON.stringify(typedData)
+  const signedTypedData = {
+    jsonrpc: '2.0',
+    method: version === 'v3' ? 'eth_signTypedData_v3' : 'eth_signTypedData',
+    params: version === 'v3' ? [txArgs.sender, jsonTypedData] : [jsonTypedData, txArgs.sender],
+    from: txArgs.sender,
+    id: new Date().getTime(),
+  }
+
+  return new Promise((resolve, reject) => {
+    web3.currentProvider.sendAsync(signedTypedData, (err, signature) => {
+      if (err) {
+        reject(err)
+        return
+      }
+
+      if (signature.result == null) {
+        reject(new Error(EIP712_NOT_SUPPORTED_ERROR_MSG))
+        return
+      }
+
+      resolve(signature.result.replace(EMPTY_DATA, ''))
```

Comment:
This is removing the hex prefix, right ... if I read the code without knowing what EMPTY_DATA is, I would not know what is going on here ;)

Reply:
how would you call `0x` prefix? it was there before I joined 😄

Reply:
I agree with Richard, maybe name it  just `0X_PREFIX` ?

Reply:
I'm questioning whether this should be a constant at all. Import statement + "0X_PREFIX" is harder to type than just `0x` :D

Reply:
If we are using that a lot of times, I think it's ok to having it on a constant, if there are just 1 or 2 places, just put `0x` in my opinion :)

Reply:
it's ok that `EMTPY_DATA` const, actually it's also used in https://github.com/gnosis/safe-react/pull/659/files#diff-2ffb1a6f81690fbba606f03ec11e4879R321 to verify if there's code for a deployed contract.

Probably, for the ease of reading/development, we should just duplicate the constant and create another with a more significant name like the one purposed by Agu or `HEX_PREFIX`.

Reply:
I'm really questioning how typing an import statement + constant name will ease the development instead of just writing `0x` but ok 😃

Reply:
so after doing some research it turns out that `0x` is used in a lot of places as a raw string, so to make it proper I'd save it for later refactoring